vnext/ReactWindowsCore/WinRTWebSocketResource.cpp

```diff
@@ -0,0 +1,438 @@
+    }
+  }
+}
+
+fire_and_forget WinRTWebSocketResource::PerformWrite()
+{
+  auto self = shared_from_this();
+  if (self->m_writeQueue.empty())
+  {
+    co_return;
+  }
+
+  try
+  {
+    co_await resume_background();
+
+    co_await resume_on_signal(self->m_connectPerformed.get());  // Ensure connection attempt has finished
+
+    co_await resume_in_queue(self->m_dispatchQueue);            // Ensure writes happen sequentially
+
+    if (self->m_readyState != ReadyState::Open)
+    {
+      self = nullptr;
+      co_return;
+    }
+
+    size_t length;
+    std::pair<string, bool> front;
+    bool popped = self->m_writeQueue.try_pop(front);
+    if (!popped)
+    {
+      throw hresult_error(E_FAIL, L"Could not retrieve outgoing message.");
+    }
+
+    auto [message, isBinary] = std::move(front);
+
+    if (isBinary)
+    {
+      self->m_socket.Control().MessageType(SocketMessageType::Binary);
+
+      auto buffer = CryptographicBuffer::DecodeFromBase64String(Utf8ToUtf16(std::move(message)));
+      length = buffer.Length();
+      self->m_writer.WriteBuffer(buffer);
+    }
+    else
+    {
+      self->m_socket.Control().MessageType(SocketMessageType::Utf8);
+
+      //TODO: Use char_t instead of uint8_t?
+      length = message.size();
+      winrt::array_view<const uint8_t> arr(
+        CheckedReinterpretCast<const uint8_t*>(message.c_str()),
+        CheckedReinterpretCast<const uint8_t*>(message.c_str()) + message.length());
+      self->m_writer.WriteBytes(arr);
+    }
+
+    co_await self->m_writer.StoreAsync();
+
+    if (self->m_writeHandler)
+    {
+      self->m_writeHandler(length);
+    }
+  }
+  catch (hresult_error const& e)
+  {
+    if (self->m_errorHandler)
+    {
+      self->m_errorHandler({ Utf16ToUtf8(e.message()), ErrorType::Ping });
+    }
+  }
+}
+
+fire_and_forget WinRTWebSocketResource::PerformClose()
+{
+  co_await resume_background();
+
+  co_await resume_on_signal(m_connectPerformed.get());
+
+  try
+  {
+    m_socket.Close(static_cast<uint16_t>(m_closeCode), Utf8ToUtf16(m_closeReason));
+
+    if (m_closeHandler)
+    {
+      m_closeHandler(m_closeCode, m_closeReason);
+    }
+  }
+  catch (winrt::hresult_invalid_argument const& e)
+  {
+    if (m_errorHandler)
+    {
+      string message;
+
+      auto restricted = e.try_as<IRestrictedErrorInfo>();
+      BSTR description;
+      HRESULT error;
+      BSTR restrictedDesc;
+      BSTR capabilitySid;
+      if (SUCCEEDED(restricted->GetErrorDetails(&description, &error, &restrictedDesc, &capabilitySid)))
+      {
+        message = Utf16ToUtf8(description) + ": " + Utf16ToUtf8(restrictedDesc);
+      }
+      else
+      {
+        message = Utf16ToUtf8(e.message());
+      }
+      // BSTR's allocated memory must be explicitly released.
+      // See https://docs.microsoft.com/en-us/cpp/atl-mfc-shared/allocating-and-releasing-memory-for-a-bstr
+      SysFreeString(description);
+      SysFreeString(restrictedDesc);
+      SysFreeString(capabilitySid);
```

Comment:
technically these are still going to be leaked if Utf16toUtf8 throws (which is possible since they use std::string). The right think would really be for these BSTRs to be in a holder object (wil has wil::unique_bstr) which calls SysFreeString in their destructor

Reply:
I recall you referencing the `wil` namespace before.
It's related to https://github.com/microsoft/wil, correct?

Is this isolated case worth adding another external code dependency?
Meaning, is there perhaps a workaround that we can write manually?

Reply:
We already have `Mso` which is like the Office version of `wil`. I would avoid adding another convenience library if possible.

@vmoroz have we brought `Mso::HandleHolder` to OSS yet?